### DISCUSSION TITLE
How to apply Edge filter?

### DISCUSSION THREAD

**Discussion by chandrakant100** [Blaze-Stars/Image-Filter-Application#29] [Q&A] [answered]
Discuss algorithm on how to apply `Edge` effect on BMP image.

**Comment by chandrakant100** [comment]
| Original Image and its Edged Image |
| ------------------------------------------- |
| <img src="https://upload.wikimedia.org/wikipedia/commons/8/85/Filtro_bordes_original_laplace_sobel.jpg" height="300px" width="700px"> |

**Comment by SHembram18** [comment]
> In this filter Highlight all of the boundaries between the object
**Lines in the image that create a boundary between one object and another.**

**Reply by chandrakant100** [reply]
please elaborate your answer.

**Reply by H1manshu21** [reply]
> > In this filter Highlight all of the boundaries between the object
> > **Lines in the image that create a boundary between one object and another.**

Hmm as @chandrakant100 can you elaborate what you wanted to say? Like tell others that how it is different from other filters or how it should be done or how should we approach it to solve it?

**Comment by Tushar7-coder** [comment]
The edge detection filters can be used to emphasize the edges and transitions in an image. 

Edge filters separate spectrum of light into two components . (i think😯 )

**Comment by SHembram18** [comment, marked as answer]
**To Implementing Edge filter on a Image:**
We have to take each pixel and modifying it based on the `3 x 3` grid of pixels that surrounds that pixel.

Algorithm : __Sobel Edge detection Algo__

- We have to use `Sobel Operator` formula ,It computes the new value of each pixel by taking weighted sum of the values for the surrounding pixels.
-  We took a pixel and formed a `3x3` square  around that pixel , to compute a new value for that same pixel , by multiplying each of the surrounding pixels by different value. ( _Values are shown in the below picture_).
-  Compute 2 weighted sums - One for detecting edges in the x & another for y direction.
-  USe this two ( __Convolutional Matrix__ )`Gx` matrix and `Gy` matrix
-  `Gx` will look for x direction boundaries and `Gy` will look for y direction boundaries. 
<img width="300" alt="sobel" src="https://user-images.githubusercontent.com/58513581/171017398-3a8cb864-d26f-478e-be3d-81c8532fb6b4.png">

**For Each Pixel**

- Compute `Gx` and `Gy` for each channel of RGB.
- For pixel which present in the border/corner, treat any pixel past the border as having all 0 values(dark black).
-  Compute each new channel value as the square root of `Gx^2 + Gy^2`

**Reply by chandrakant100** [reply]
Good Approach !!

**Comment by chandrakant100** [comment]
## Pseudocode for Edge Filter
```
function edges (Argument1 rows, Argument2 columns, Argument3 pixel 2Darray ) {
      create initial variables to store values of the total sum in each pixel channel for both the Gx and Gy matrix.
      e.g ( sumBluex, sumRedx, sumGreenx, sumBluey, sumRedy, sumGreeny)
      create a temporary 2DPixel array to store the results of the calculation.
       iterate and pick each row of the pixel 2D array
                iterate and pick each pixel of the row selected above
                   initialize previously created variables with a default value of 0.
                   generate a loop to check the neighboring pixel in a 3x3 grid.
                  sums values of the pixel and 8 neighboring ones after applying a modifier, skips iteration if 
                   it goes outside the picture.
                   calculate modifiers for vertical and horizontal borders.
                   sum up them with the resultant pixel grid and update each channel of the temporary 2D 
                   array, with values of each pixel channel (i.e red, green, blue )will be sqrt(Gx^2 + Gy^2).
                   If any value exceeds 255 round it to exact 255 and stores it in the channel.
                   After the Whole iteration is completed, copy the values of the Temp 2D array into the main 
                   Pixel 2D array.
}
```
